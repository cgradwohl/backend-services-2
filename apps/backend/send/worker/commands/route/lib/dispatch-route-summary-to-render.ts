import { renderService } from "~/send/service";
import { IRenderProviderPayload } from "~/send/types";
import { IRoutingSummary } from "../types";

export async function dispatchRouteSummaryToRenderService({
  contextFilePath,
  dryRunKey,
  messageFilePath,
  messageId,
  requestId,
  routingSummary,
  tenantId,
  shouldVerifyRequestTranslation = false,
  translated = false,
}: {
  contextFilePath: string;
  dryRunKey: string;
  messageFilePath: string;
  messageId: string;
  requestId: string;
  routingSummary: Partial<IRoutingSummary>[];
  tenantId: string;
  shouldVerifyRequestTranslation?: boolean;
  translated?: boolean;
}) {
  const dispatches = routingSummary
    .filter((summary) => summary.selected)
    .map(
      (summary): IRenderProviderPayload => ({
        command: "render",
        channel: summary.channel,
        channelId: summary.id!,
        contextFilePath: contextFilePath,
        dryRunKey,
        messageId,
        messageFilePath,
        configurationId: summary.configurationId!,
        requestId,
        tenantId,
        shouldVerifyRequestTranslation,
        translated,
      })
    )
    .map((payload) => renderService(tenantId).emit(payload));

  await Promise.all(dispatches);
}
