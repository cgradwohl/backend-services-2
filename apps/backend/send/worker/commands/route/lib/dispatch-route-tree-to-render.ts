import { createUnroutableEvent } from "~/lib/dynamo/event-logs";
import {
  RouteLeaf,
  RouteNode,
  RouteNodeAddress,
  SendTimes,
  setSendTimesForLeafs,
  getAllRouteLeafs,
  getFailoverRouteNode,
  RouteTimeoutTable,
  getTimedOutLeafs,
  GetFailoverNodeResult,
} from "~/lib/send-routing";
import { renderService } from "~/send/service";
import { IRenderProviderPayload } from "~/send/types";

export async function dispatchRouteTreeToRenderService({
  contextFilePath,
  dryRunKey,
  messageFilePath,
  messageId,
  requestId,
  routingTree,
  tenantId,
  times: baseTimes,
  failedAddress,
  timeouts,
}: {
  contextFilePath: string;
  dryRunKey: string;
  messageFilePath: string;
  messageId: string;
  requestId: string;
  routingTree: RouteNode;
  tenantId: string;
  times?: SendTimes;
  failedAddress?: RouteNodeAddress;
  timeouts?: RouteTimeoutTable;
}) {
  const branch = getBranch({
    tree: routingTree,
    failedAddress,
    times: baseTimes,
    timeouts,
  });

  if (typeof branch === "string") {
    await handleFailoverResult({ result: branch, tenantId, messageId });
    return;
  }

  // We send to all leafs of the current branch (Remember, branches are like a recursive send to all)
  const leafs = getAllRouteLeafs(branch);

  // See [1]
  const times = setSendTimesForLeafs(leafs, baseTimes);

  const leafMapper = (leaf: RouteLeaf): IRenderProviderPayload => ({
    command: "render",
    channel: leaf.channel,
    channelId: leaf.templateChannelId,
    contextFilePath: contextFilePath,
    dryRunKey,
    messageId,
    messageFilePath,
    configurationId: leaf.providerConfigurationId!,
    requestId,
    tenantId,
    address: leaf.address,
    times,
    shouldVerifyRequestTranslation: false,
    translated: false,
  });

  const dispatches = leafs
    .map(leafMapper)
    .map((payload) => renderService(tenantId).emit(payload));

  await Promise.all(dispatches);
}

function getBranch(opts: {
  tree: RouteNode;
  failedAddress?: RouteNodeAddress;
  times?: SendTimes;
  timeouts?: RouteTimeoutTable;
}): GetFailoverNodeResult {
  const { tree, failedAddress, times, timeouts } = opts;
  if (!failedAddress || !times || !timeouts) {
    return tree;
  }

  const allFailedNodes = [
    failedAddress,
    ...getTimedOutLeafs({ tree, times, timeouts }).map((l) => l.address),
  ];

  return getFailoverRouteNode({ tree, failedAddress, allFailedNodes });
}

async function handleFailoverResult({
  result,
  tenantId,
  messageId,
}: {
  result: GetFailoverNodeResult;
  tenantId: string;
  messageId: string;
}) {
  if (typeof result !== "string") {
    return;
  }

  if (result === "failover-strategies-exhausted") {
    await createUnroutableEvent(
      tenantId,
      messageId,
      "FAILED",
      "All failover options for send have been exhausted."
    );
  }

  if (result === "sibling-routes-may-not-have-failed") {
    await createUnroutableEvent(
      tenantId,
      messageId,
      "BOUNCED",
      "Cannot failover when status of sibling provider sends is unknown."
    );
  }
}

// [1] The times table is inherently forked each time route is invoked. This means that
// two independent failover strategies will create two times tables.
