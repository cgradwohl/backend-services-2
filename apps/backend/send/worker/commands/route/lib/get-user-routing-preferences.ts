import { ISendMessageContext } from "~/send/types";
import {
  getUserPreference,
  mapPreferences,
  PreferenceUndeliveryReason,
} from "~/lib/preferences";
import { INotificationWire } from "~/types.api";
import isNotificationWire from "~/send/utils/is-notification-wire";
import {
  pluckPreferenceRulesByTypes,
  ruleHandlers,
  PreferenceRules,
} from "~/preferences/rules";

interface IRoutingLogByPreferences {
  type: PreferenceUndeliveryReason;
  reason: string;
  data: Record<string, unknown>;
}

/** Determines if user is subscribed to notification. Does not handle channel preferences */
export function getUserRoutingPreferences(
  context: ISendMessageContext
): IRoutingLogByPreferences | undefined {
  const preferences = mapPreferences(context?.preferences);
  const category = context?.category;
  const content = context.content;

  const notification = (isNotificationWire(content) &&
    content) as Partial<INotificationWire>;

  const userPreference = getUserPreference({
    category: context.category,
    notification: context.content as INotificationWire,
    preferences: context.preferences!,
    event: context.metadata?.event,
  });

  if (userPreference) {
    return {
      data: mapPreferences(preferences),
      reason: userPreference.message,
      type: userPreference.reason,
    };
  }

  const [categoryPreferenceRules, notificationPreferenceRules] =
    pluckPreferenceRulesByTypes(category!, notification, preferences);

  const [hasSnoozedCategory] = categoryPreferenceRules
    .filter((rule) => rule.type === "snooze")
    .map((rule) => (ruleHandlers as PreferenceRules)[rule.type](rule));

  const [hasSnoozedNotification] = notificationPreferenceRules
    .filter((rule) => rule.type === "snooze")
    .map((rule) => (ruleHandlers as PreferenceRules)[rule.type](rule));

  if (hasSnoozedCategory || hasSnoozedNotification) {
    const reason = hasSnoozedCategory
      ? "Snoozed at category level by user"
      : "Snoozed at notification level by user";

    return {
      data: {
        ...(categoryPreferenceRules !== null && { categoryPreferenceRules }),
        ...(notificationPreferenceRules !== null && {
          notificationPreferenceRules,
        }),
      },
      reason,
      type: "UNSUBSCRIBED",
    };
  }
}
