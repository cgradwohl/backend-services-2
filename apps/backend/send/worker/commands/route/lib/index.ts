export { ChannelHandleFailedError } from "~/lib/send-routing";
export * from "./get-user-routing-preferences";
export * from "./dispatch-route-summary-to-render";
