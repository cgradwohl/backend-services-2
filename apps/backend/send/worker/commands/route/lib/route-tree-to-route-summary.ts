import {
  RouteNode,
  getAllRouteLeafs,
  RoutingSummary,
} from "~/lib/send-routing";

export const routeTreeToRouteSummary = (tree?: RouteNode): RoutingSummary[] => {
  if (!tree) return [];

  return getAllRouteLeafs(tree, { includeDeadLeafs: true }).map(
    (leaf): RoutingSummary => {
      if (leaf.type === "dead-leaf") {
        return {
          channel: leaf.channel,
          provider: leaf.provider,
          reason: leaf.failureReason,
          type: leaf.failureType,
          selected: false,
        };
      }

      if (leaf.type === "dead-branch") {
        return {
          channel: leaf.channel,
          reason: leaf.failureReason,
          type: leaf.failureType,
          selected: false,
        };
      }

      return {
        channel: leaf.channel,
        provider: leaf.provider,
        taxonomy: leaf.taxonomy,
        configurationId: leaf.providerConfigurationId,
        selected: true,
        id: leaf.templateChannelId,
      };
    }
  );
};
