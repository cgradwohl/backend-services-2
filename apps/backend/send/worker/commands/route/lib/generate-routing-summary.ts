import { VM } from "vm2";
import { MessageChannels } from "~/api/send/types";
import { pipe } from "~/lib/pipe";
import { RoutingStrategy, RoutingSummary } from "~/lib/send-routing/types";
import {
  getChannelsByProvider,
  getChannelByProvider,
  getClassByProvider,
  getTaxonomyFromProvider,
} from "~/lib/taxonomy-helpers";
import { TokensByProvider } from "~/lib/token-storage";
import providers from "~/providers";
import { IProviderConfiguration } from "~/send/types";
import { INotificationWire } from "~/types.api";

/**
 * @deprecated
 * TODO: Remove once the july-2022-routing-tree-enabled has been enabled for everyone and run without issue
 */
export const generateRoutingSummary = async ({
  providerConfigs,
  params,
  strategy,
  notification,
  tokens,
}: {
  providerConfigs: IProviderConfiguration[];
  params: Record<"data" | "profile", any>;
  notification?: INotificationWire;
  strategy: RoutingStrategy;
  tokens?: TokensByProvider;
}): Promise<RoutingSummary[]> => {
  const filteredProviderConfigs = filterProviderConfigs({
    params,
    providerConfigs,
    strategy,
  });

  const summary = await walkRouting({
    params,
    providerConfigs: filteredProviderConfigs,
    strategy,
    tokens,
  });

  return pipe(summary)
    .into((summary) => applyNotificationToRoutingSummary(summary, notification))
    .complete();
};

export const walkRouting = async ({
  params,
  providerConfigs,
  strategy,
  tokens,
}: {
  params: Record<"data" | "profile", any>;
  providerConfigs: IProviderConfiguration[];
  strategy: RoutingStrategy;
  tokens?: TokensByProvider;
}): Promise<RoutingSummary[]> => {
  const {
    routing,
    channels: messageChannels,
    providers: messageProviders,
  } = strategy;

  const summary: RoutingSummary[] = [];
  for (const channel of routing.channels) {
    if (typeof channel === "string") {
      summary.push(
        ...(await generateSummaryFromChannel({
          channel,
          messageChannels,
          params,
          providerConfigs,
          tokens,
        }))
      );
    }

    if (typeof channel === "object") {
      summary.push(
        ...(await walkRouting({
          strategy: {
            routing: channel,
            channels: messageChannels,
            providers: messageProviders,
          },
          params,
          providerConfigs,
          tokens,
        }))
      );
    }

    if (summary.some((res) => res.selected) && routing.method === "single") {
      break;
    }
  }

  return summary;
};

export const generateSummaryFromChannel = async ({
  channel,
  messageChannels,
  params,
  providerConfigs,
  tokens,
}: {
  channel: string;
  messageChannels: MessageChannels;
  params: Record<"data" | "profile", any>;
  providerConfigs: IProviderConfiguration[];
  tokens?: TokensByProvider;
}): Promise<RoutingSummary[]> => {
  const messageChannel = messageChannels[channel];

  if (!evaluateStrategyConditional({ condition: messageChannel?.if, params })) {
    return [
      {
        channel,
        type: "FILTERED",
        selected: false,
        provider: "",
        reason: "Channel conditional failed",
      },
    ];
  }

  const providerCandidates = findConfiguredProvidersForChannel({
    channel,
    providerConfigs,
  }).filter(({ providerKey }) =>
    configuredProviderAllowedByChannel({
      providerKey,
      messageChannelProviders: messageChannel?.providers,
    })
  );

  if (providerCandidates.length === 0) {
    return [
      {
        channel,
        type: "NO_PROVIDERS",
        selected: false,
        reason:
          "No remaining configured providers for channel or channel is invalid", // Provider may have failed a conditional
        provider: "",
      },
    ];
  }

  return generateSummaryFromProviders({
    channel,
    params,
    providerConfigs: providerCandidates,
    routingMethod: messageChannel?.routing_method,
    tokens,
  });
};

export const generateSummaryFromProviders = async ({
  channel,
  params,
  providerConfigs,
  routingMethod = "single",
  tokens,
}: {
  channel: string;
  params: Record<"data" | "profile", any>;
  providerConfigs: ConfiguredProvider[];
  routingMethod?: "all" | "single";
  tokens?: TokensByProvider;
}): Promise<RoutingSummary[]> => {
  const summary: RoutingSummary[] = [];
  for (const provider of providerConfigs) {
    const { data, profile } = params;
    const canHandleResult = providers[provider.providerKey]?.handles({
      config: provider.config,
      data,
      profile,
      tokensByProvider: tokens,
    });

    const canHandle: boolean | Error =
      canHandleResult instanceof Promise
        ? await canHandleResult.catch((e) => e)
        : canHandleResult;

    if (canHandle instanceof Error) {
      summary.push({
        channel,
        provider: provider.providerKey,
        type: "PROVIDER_ERROR",
        reason:
          "Provider returned an error when checking its ability to handle the notification",
        selected: false,
      });
      continue;
    }

    if (!canHandle) {
      summary.push({
        channel,
        provider: provider.providerKey,
        type: "MISSING_PROVIDER_SUPPORT",
        reason: `Message is missing the minimum data required to send with this provider. Please check the message's profile and data properties and try again.`,
        selected: false,
      });
      continue;
    }

    summary.push(configuredProviderToSummary(channel, provider));
    if (routingMethod === "single") {
      break;
    }
  }
  return summary;
};

export const evaluateStrategyConditional = ({
  condition,
  params,
  type = "Channel",
}: {
  condition?: string;
  params: Record<"data" | "profile", any>;
  type?: "Channel" | "Provider";
}): boolean => {
  if (!condition) {
    return true;
  }

  const { data, profile } = params;

  const vm = new VM({ sandbox: { data, profile } });
  const result = vm.run(condition);

  if (typeof result !== "boolean") {
    throw new Error(
      `${type} conditional does not evaluate to boolean: ${condition}`
    );
  }

  return result;
};

export const configuredProviderToSummary = (
  channel: string,
  provider: ConfiguredProvider
): RoutingSummary => ({
  channel,
  provider: provider.providerKey,
  selected: true,
  configurationId: provider.config.id,
  taxonomy: provider.taxonomy,
});

export const findConfiguredProvidersForChannel = ({
  channel,
  providerConfigs,
}: {
  channel: string;
  providerConfigs: IProviderConfiguration[];
}): ConfiguredProvider[] =>
  providerConfigs
    .filter((config) => {
      const providerKey = config.json.provider;
      return (
        getChannelsByProvider(providerKey).includes(channel) ||
        getChannelByProvider(providerKey) === channel ||
        getClassByProvider(providerKey) === channel ||
        providerKey === channel
      );
    })
    .map((config) => ({
      config,
      providerKey: config.json.provider,
      taxonomy: getTaxonomyFromProvider(config.json.provider),
    }));

export const configuredProviderAllowedByChannel = ({
  providerKey,
  messageChannelProviders,
}: {
  providerKey: string;
  messageChannelProviders?: string[];
}): boolean => {
  if (!messageChannelProviders) {
    return true;
  }

  return messageChannelProviders.includes(providerKey);
};

type ConfiguredProvider = {
  providerKey: string;
  config: IProviderConfiguration;
  taxonomy: string;
};

export const filterProviderConfigs = ({
  params,
  providerConfigs,
  strategy,
}: {
  params: Record<"data" | "profile", any>;
  providerConfigs: IProviderConfiguration[];
  strategy: RoutingStrategy;
}): IProviderConfiguration[] =>
  providerConfigs.filter((config) => {
    const providerKey = config.json.provider;
    const messageProvider = strategy.providers[providerKey];
    return evaluateStrategyConditional({
      condition: messageProvider?.if,
      params,
      type: "Provider",
    });
  });

export const applyNotificationToRoutingSummary = (
  summary: RoutingSummary[],
  notification?: INotificationWire
): RoutingSummary[] => {
  if (!notification) {
    return summary;
  }

  const iChannels = [
    ...(notification.json?.channels?.always ?? []),
    ...(notification.json?.channels?.bestOf ?? []),
  ];

  return summary.map((result): RoutingSummary => {
    const iChannel = iChannels.find((channel) =>
      channel.taxonomy.includes(result.channel)
    );
    if (!iChannel) {
      return result;
    }

    return {
      ...result,
      id: iChannel.id,
    };
  });
};
