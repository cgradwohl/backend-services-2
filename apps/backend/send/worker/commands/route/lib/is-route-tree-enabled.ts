import { findPricingPlanForTenant } from "~/lib/plan-pricing";
import { ITenant } from "~/types.api";

export const isRouteTreeEnabled = (
  tenant: ITenant,
  shouldUseRouteTree?: boolean
) => {
  return findPricingPlanForTenant(tenant) === "custom" || shouldUseRouteTree;
};
