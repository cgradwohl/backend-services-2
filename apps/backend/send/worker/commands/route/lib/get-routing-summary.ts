import { ISendMessageContext } from "~/send/types";
import isNotificationWire from "~/send/utils/is-notification-wire";
import { IRoutingSummary } from "../types";
import { generateRoutingSummary } from "./generate-routing-summary";
import { getTemplateV1RoutingSummary } from "~/lib/send-routing";

/** @deprecated Remove after failover hits GA */
export async function getRoutingSummary(
  context: ISendMessageContext
): Promise<Partial<IRoutingSummary>[]> {
  if (!!context.strategy) {
    return generateRoutingSummary({
      providerConfigs: context.providers,
      strategy: context.strategy,
      notification: isNotificationWire(context.content)
        ? context.content
        : undefined,
      params: {
        data: context.data,
        profile: context.profile,
      },
      tokens: context.variableData?.tokens,
    });
  }

  if (isNotificationWire(context.content)) {
    const result = await getTemplateV1RoutingSummary({
      category: context.category,
      templateV1: context.content,
      preferences: context.preferences,
      providers: context.providers,
      variableData: context.variableData,
      tokens: context.variableData?.tokens,
    });
    return [...result.always, ...result.bestOf];
  }

  return [];
}
