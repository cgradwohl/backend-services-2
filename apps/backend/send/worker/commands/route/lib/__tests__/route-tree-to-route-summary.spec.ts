import { RouteNode } from "~/lib/send-routing";
import { routeTreeToRouteSummary } from "../route-tree-to-route-summary";

describe("route tree to route summary", () => {
  it("should emit a well formed summary", () => {
    expect(routeTreeToRouteSummary(routing)).toMatchSnapshot();
  });
});

const routing: RouteNode = {
  address: [],
  nodes: [
    {
      address: [0],
      nodes: [
        {
          channel: "sms",
          provider: "twilio",
          providerConfigurationId: "7593c6b6-f241-47f0-85b4-b5386cd60086",
          taxonomy: "direct_message:sms:twilio",
          address: [0, 0],
          providerFailoverIndex: 1,
          type: "leaf",
        },
      ],
      type: "branch",
      failover: {
        address: [0, "failover"],
        nodes: [
          {
            channel: "sms",
            provider: "vonage",
            providerConfigurationId: "7593c6b6-f241-47f0-85b4-vonage",
            taxonomy: "direct_message:sms:vonage",
            address: [0, "failover", 0],
            providerFailoverIndex: 2,
            type: "leaf",
          },
        ],
        type: "branch",
      },
    },
    {
      channel: "email",
      failureType: "FILTERED",
      failureReason: "Channel conditional failed",
      address: [1],
      type: "dead-branch",
    },
  ],
  type: "branch",
};
