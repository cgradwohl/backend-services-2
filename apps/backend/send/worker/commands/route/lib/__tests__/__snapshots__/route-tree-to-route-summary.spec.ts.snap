// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`route tree to route summary should emit a well formed summary 1`] = `
Array [
  Object {
    "channel": "sms",
    "configurationId": "7593c6b6-f241-47f0-85b4-b5386cd60086",
    "id": undefined,
    "provider": "twilio",
    "selected": true,
    "taxonomy": "direct_message:sms:twilio",
  },
  Object {
    "channel": "email",
    "reason": "Channel conditional failed",
    "selected": false,
    "type": "FILTERED",
  },
]
`;
